extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java

```diff
@@ -98,220 +97,207 @@ public String fromBytes(byte[] bytes)
           }
         }
       }
-      catch (InterruptedException | ExecutionException | TimeoutException e) {
-        if (!future.isDone() && !future.cancel(true) && !future.isDone()) {
-          LOG.warn("Could not cancel kafka listening thread");
-        }
-        LOG.error(e, "Failed to start kafka extraction factory");
-        cacheManager.delete(factoryId);
-        return false;
-      }
+    };
+  }
 
-      started.set(true);
-      return true;
-    }
+  void onNewEntry(String key, String message) {
+    doubleEventCount.incrementAndGet();
+    map.put(key, message);
+    doubleEventCount.incrementAndGet();
+    LOG.trace("Placed key[%s] val[%s]", key, message);
   }
 
-  // Overriden in tests
-  ConsumerConnector buildConnector(Properties properties)
-  {
-    return new kafka.javaapi.consumer.ZookeeperConsumerConnector(
-        new ConsumerConfig(properties)
+  private Properties buildProperties() {
+    final Properties kafkaProperties = new Properties();
+    kafkaProperties.putAll(this.kafkaProperties);
+    if (kafkaProperties.containsKey("group.id")) {
+      throw new IAE(
+          "Cannot set kafka property [group.id]. Property is randomly generated for you. Found [%s]",
+          kafkaProperties.getProperty("group.id")
+      );
+    }
+    if (kafkaProperties.containsKey("auto.offset.reset")) {
+      throw new IAE(
+          "Cannot set kafka property [auto.offset.reset]. Property will be forced to [smallest]. Found [%s]",
+          kafkaProperties.getProperty("auto.offset.reset")
+      );
+    }
+    Preconditions.checkNotNull(
+        kafkaProperties.getProperty("zookeeper.connect"),
+        "zookeeper.connect required property"
     );
+
+    kafkaProperties.setProperty("group.id", factoryId);
+
+    // Enable publish-subscribe
+    kafkaProperties.setProperty("auto.offset.reset", "smallest");
+    return kafkaProperties;
   }
 
+
   @Override
   public boolean close()
   {
-    synchronized (started) {
-      if (!started.get() || executorService.isShutdown()) {
-        LOG.info("Already shutdown, ignoring");
-        return !started.get();
-      }
-      started.set(false);
-      executorService.shutdownNow();
-      final ListenableFuture<?> future = this.future;
-      if (future != null) {
-        if (!future.isDone() && !future.cancel(true) && !future.isDone()) {
-          LOG.error("Error cancelling future for topic [%s]", getKafkaTopic());
-          return false;
-        }
-      }
-      if (!cacheManager.delete(factoryId)) {
-        LOG.error("Error removing [%s] for topic [%s] from cache", factoryId, getKafkaTopic());
+    if(!advanceState(State.RUNNING, State.STOPPING)) {
+      LOG.info("Already shutdown, ignoring. Current state: [%s]", currentState);
+      return currentState.get().equals(State.TERMINATED);
+    }
+
+    // TODO: if current state is STOPPING, block until a terminal state is reached?
+    // Is this method ever invoked by multiple threads simultaneously?
```

Comment:
What is the expected behaviour if `close()` is invoked by multiple threads simultaneously?